hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/steps/StakePeriodChanges.java

```diff
@@ -172,15 +179,14 @@ public static boolean isNextStakingPeriod(
         }
     }
 
-    private void keyNewRoster(
+    private void startKeyingCandidateRoster(
             @NonNull final HandleContext handleContext, @NonNull final WritableRosterStore rosterStore) {
         final var nodeStore = handleContext.storeFactory().readableStore(ReadableNodeStore.class);
-        final var newCandidateRoster = nodeStore.newRosterFromNodes();
-
-        if (!Objects.equals(newCandidateRoster, rosterStore.getCandidateRoster())
-                && !Objects.equals(newCandidateRoster, rosterStore.getActiveRoster())) {
-            rosterStore.putCandidateRoster(newCandidateRoster);
-            tssBaseService.setCandidateRoster(newCandidateRoster, handleContext);
+        final var roster = nodeStore.snapshotOfFutureRoster();
+        if (!Objects.equals(roster, rosterStore.getCandidateRoster())
+                && !Objects.equals(roster, rosterStore.getActiveRoster())) {
+            rosterStore.putCandidateRoster(roster);
+            tssBaseService.setCandidateRoster(roster, handleContext);
```

Comment:
I'm wondering if the first operation should trigger the second one automatically? Also, I'm wondering if the TSS Base Service is supposed to read the actual candidate roster from the roster store instead of receiving a random roster as an input argument?

The code above doesn't seem robust enough as one can forget to rekey a new candidate roster if they miss the second call, or they can initiate a rekeying of a roster that isn't a candidate (or, in an edge case of a new network, the current active) roster.

Reply:
Agreed Anthony, this doesn't seem like _quite_ the right place to orchestrate things...will keep thinking about this. 🤔